File: tools/verify_warranty_status_n8n_tool.py
"""n8n Custom Code Tool: mock warranty status lookup."""

import json
from calendar import monthrange
from datetime import date, datetime


MOCK_WARRANTY_LOOKUP = {
    "GTW465ASNWW": {
        "coverage_months": 12,
        "coverage_type": "standard laundry coverage",
        "notes": "Parts and labor are covered for the first year.",
    },
    "GFW850SPNRS": {
        "coverage_months": 36,
        "coverage_type": "premium front-load washer coverage",
        "notes": "Includes motor and drum coverage through the extended term.",
    },
    "GSS25GYPFS": {
        "coverage_months": 24,
        "coverage_type": "refrigerator sealed-system coverage",
        "notes": "Sealed-system components receive extra attention in service review.",
    },
    "JGB735SPSS": {
        "coverage_months": 18,
        "coverage_type": "gas range limited coverage",
        "notes": "Igniter and control-board diagnostics are prioritized.",
    },
    "PDT715SYNFS": {
        "coverage_months": 30,
        "coverage_type": "dishwasher quiet-package coverage",
        "notes": "Pump, rack, and leak-sensor claims are eligible during coverage.",
    },
}


def parse_input(raw_query):
    if isinstance(raw_query, dict):
        return raw_query

    try:
        return json.loads(str(raw_query))
    except Exception:
        return {}


def verify_warranty_status(model_name, purchase_date):
    normalized_model = str(model_name or "").strip().upper()

    if not normalized_model or not purchase_date:
        return {
            "model_name": normalized_model,
            "status": "missing_input",
            "coverage_end_date": None,
            "days_remaining": None,
            "details": "Provide model and purchase_date.",
        }

    warranty_record = MOCK_WARRANTY_LOOKUP.get(normalized_model)

    if warranty_record is None:
        return {
            "model_name": normalized_model,
            "status": "unknown_model",
            "coverage_end_date": None,
            "days_remaining": None,
            "details": "No mock warranty rule is available for this model.",
        }

    try:
        purchased_on = datetime.strptime(str(purchase_date), "%Y-%m-%d").date()
    except ValueError:
        return {
            "model_name": normalized_model,
            "status": "invalid_purchase_date",
            "coverage_end_date": None,
            "days_remaining": None,
            "details": "Use YYYY-MM-DD format for purchase_date.",
        }

    coverage_months = int(warranty_record["coverage_months"])
    month_index = purchased_on.month - 1 + coverage_months
    coverage_year = purchased_on.year + month_index // 12
    coverage_month = month_index % 12 + 1
    coverage_day = min(purchased_on.day, monthrange(coverage_year, coverage_month)[1])
    coverage_end = date(coverage_year, coverage_month, coverage_day)
    days_remaining = (coverage_end - date.today()).days
    status = "active" if days_remaining >= 0 else "expired"

    return {
        "model_name": normalized_model,
        "status": status,
        "coverage_end_date": coverage_end.isoformat(),
        "days_remaining": max(days_remaining, 0),
        "details": f"{warranty_record['coverage_type']}: {warranty_record['notes']}",
    }


try:
    raw_query = _query
except NameError:
    try:
        raw_query = query
    except NameError:
        raw_query = "{}"
tool_input = parse_input(raw_query)
result = verify_warranty_status(
    tool_input.get("model"),
    tool_input.get("purchase_date"),
)

return json.dumps(result)

File: tools/verify_warranty_status_n8n_tool.js
"use strict";

const MOCK_WARRANTY_LOOKUP = {
  GTW465ASNWW: {
    coverageMonths: 12,
    coverageType: "standard laundry coverage",
    notes: "Parts and labor are covered for the first year.",
  },
  GFW850SPNRS: {
    coverageMonths: 36,
    coverageType: "premium front-load washer coverage",
    notes: "Includes motor and drum coverage through the extended term.",
  },
  GSS25GYPFS: {
    coverageMonths: 24,
    coverageType: "refrigerator sealed-system coverage",
    notes: "Sealed-system components receive extra attention in service review.",
  },
  JGB735SPSS: {
    coverageMonths: 18,
    coverageType: "gas range limited coverage",
    notes: "Igniter and control-board diagnostics are prioritized.",
  },
  PDT715SYNFS: {
    coverageMonths: 30,
    coverageType: "dishwasher quiet-package coverage",
    notes: "Pump, rack, and leak-sensor claims are eligible during coverage.",
  },
};

const MS_PER_DAY = 24 * 60 * 60 * 1000;

function parseInput(rawQuery) {
  if (typeof rawQuery === "object" && rawQuery !== null) {
    return rawQuery;
  }

  try {
    return JSON.parse(String(rawQuery));
  } catch {
    return {};
  }
}

function parseIsoDate(value) {
  const match = /^(\d{4})-(\d{2})-(\d{2})$/.exec(String(value));

  if (!match) {
    return null;
  }

  const year = Number(match[1]);
  const month = Number(match[2]);
  const day = Number(match[3]);
  const parsedDate = new Date(year, month - 1, day);

  if (
    parsedDate.getFullYear() !== year ||
    parsedDate.getMonth() !== month - 1 ||
    parsedDate.getDate() !== day
  ) {
    return null;
  }

  return parsedDate;
}

function addMonthsClamped(date, monthsToAdd) {
  const monthIndex = date.getMonth() + monthsToAdd;
  const coverageYear = date.getFullYear() + Math.floor(monthIndex / 12);
  const coverageMonth = monthIndex % 12;
  const lastDayOfCoverageMonth = new Date(
    coverageYear,
    coverageMonth + 1,
    0,
  ).getDate();
  const coverageDay = Math.min(date.getDate(), lastDayOfCoverageMonth);

  return new Date(coverageYear, coverageMonth, coverageDay);
}

function startOfToday() {
  const now = new Date();

  return new Date(now.getFullYear(), now.getMonth(), now.getDate());
}

function verifyWarrantyStatus(modelName, purchaseDate) {
  const normalizedModel = String(modelName || "").trim().toUpperCase();

  if (!normalizedModel || !purchaseDate) {
    return {
      model_name: normalizedModel,
      status: "missing_input",
      coverage_end_date: null,
      days_remaining: null,
      details: "Provide model and purchase_date.",
    };
  }

  const warrantyRecord = MOCK_WARRANTY_LOOKUP[normalizedModel];

  if (!warrantyRecord) {
    return {
      model_name: normalizedModel,
      status: "unknown_model",
      coverage_end_date: null,
      days_remaining: null,
      details: "No mock warranty rule is available for this model.",
    };
  }

  const purchasedOn = parseIsoDate(purchaseDate);

  if (!purchasedOn) {
    return {
      model_name: normalizedModel,
      status: "invalid_purchase_date",
      coverage_end_date: null,
      days_remaining: null,
      details: "Use YYYY-MM-DD format for purchase_date.",
    };
  }

  const coverageEnd = addMonthsClamped(
    purchasedOn,
    warrantyRecord.coverageMonths,
  );
  const daysRemaining = Math.round((coverageEnd - startOfToday()) / MS_PER_DAY);
  const status = daysRemaining >= 0 ? "active" : "expired";

  return {
    model_name: normalizedModel,
    status,
    coverage_end_date: coverageEnd.toISOString().slice(0, 10),
    days_remaining: Math.max(daysRemaining, 0),
    details: `${warrantyRecord.coverageType}: ${warrantyRecord.notes}`,
  };
}

const input = parseInput(query);
const result = verifyWarrantyStatus(input.model, input.purchase_date);

return JSON.stringify(result);
